### ISSUE TITLE
useQueryLoader() hook to handle default loading and error rendering

### ISSUE THREAD

**Issue by mbrowne** [apollographql/apollo-feature-requests#106]
In general, react-apollo is great for running GraphQL queries with minimal boilerplate, but one thing that gets quite repetitive is handling loading and error states (always checking if (loading), etc.) In the apps I have worked on, usually you want the same loading spinner and same error rendering at least 90% of the time. In order to handle this typical case, I previously created a `<QueryLoader>` component, which I have now refactored into a hook, `useQueryLoader()`.

Example usage:
```js
const App = () => (
  <ApolloProvider
    client={client}
    options={{
      defaultLoadingComponent: Loader,
      defaultErrorComponent: ErrorMessage
    }}
  >
    <Dogs />
  </ApolloProvider>
);
...
const Dogs = () => {
  return useQueryLoader(GET_DOGS)(({ data }) => (
    <ul>
      {data.dogs.map(dog => (
        <li key={dog.id}>{dog.breed}</li>
      ))}
    </ul>
  ));
};
```

For a more complete example, see my PR.

## Including with react-apollo

You might ask, why should this hook be included with react-apollo, instead of just being a separate library or left to each app to implement? The main advantage of having it included with react-apollo, aside from convenience, is that it would be more difficult or less elegant to configure the default loading and error components without having a built-in way to attach them to the `<ApolloProvider>`. They'd probably either have to be static properties (which wouldn't allow configuring different default components for different providers), or it would be necessary to create an additional provider just for that purpose.

I realize that bundle size is important, so I understand if this feature request is rejected for that reason. But I think this would be a convenient and popular feature if it were included that fits in nicely with Apollo's design goals with respect to usability.

**Comment by mbrowne** [comment]
Here is my PR implementing this:
https://github.com/apollographql/react-apollo/pull/3079

**Comment by FredyC** [comment]
Every project has different needs regarding how to tackle loading and error states. I doubt it's possible to invent one universal way to tackle or scenarios gracefully.

Besides, when Suspense for data fetching is out, it becomes irrelevant.

**Comment by mbrowne** [comment]
I could certainly turn this into a separate library (which would require its own Context provider) if not enough Apollo users would use it to make it worth including I'm react-apollo.

I disagree on the second point. The `queryLoader` hook could easily be updated to use suspense internally when it's fully ready. In fact I did an earlier implementation of this hook using the third-party react-apollo-hooks library using suspense.

**Comment by FredyC** [comment]
> I disagree on the second point. The `queryLoader` hook could easily be updated to use suspense internally when it's fully ready.

You are clearly confused here. You have used Suspense which was never intended for data fetching scenarios. We can't really tell what it will look like in the end at this point. How can you be so sure it will be easily adaptable?

**Comment by mbrowne** [comment]
@FredyC (replying late) This is a moot point now as far as react-apollo itself is concerned (my PR was rejected due to very understandable concerns about bundle sizes of the core packages), but I figured I'd reply anyway in case anyone else is interested in using a similar pattern...

While certain things about suspense are still up in the air, I think this comment is a pretty good summary of the situation:
https://github.com/apollographql/react-apollo/issues/2539#issuecomment-465105605

The basic pattern of Suspense throwing a promise and having a fallback element is very unlikely to change, and that would all be handled inside the `useQueryLoader()` hook anyhow, so its implementation could easily change as needed.

**Comment by FredyC** [comment]
Right...However Suspense turns out, I am still not convinced this is a good pattern. Just seeing codebase of a single medium-sized app and there are some differences on how loading needs to be handled.

Besides, you are assuming, that all cases are simple "call query & render result", but in practice, there is usually a bunch of other logic in between. For example, filtering data based on app state. What about `useMemo` for such filtered results? Or even `useEffect` that needs to do something else over the data. You cannot have that inside another hook.

Some example from a real app (+some MobX magic, not important). We have generated a hook for each query. It would be very inconvenient doing it some other way around. As you can see the loading state is expressed very specifically here.

```tsx
export const LogoutWidget = observer(() => {
  const { auth } = useRoot()
  const { data, loading } = useQLogoutWidget(
    { id: auth.userId },
    { skip: !auth.isAuthenticated },
  )

  let displayName
  if (loading) {
    displayName = i18n.t`Not logged in`
  } else {
    displayName = `${data.user.firstName} ${data.user.lastName}`
  }

  const now = useTimer({ enabled: true, interval: SECOND })

  return (
    <StyledLogoutWidget data-testid="logout-widget">
      <Link to="/logout">
        <span className="name">{displayName}</span>
        <ExitToApp />
      </Link>
      <div className="time">{DF.format(now, 'H:mm:ss')}</div>
    </StyledLogoutWidget>
  )
})
```

**Comment by mbrowne** [comment]
Couldn't you just use `useQuery` directly for such cases? I would never suggest that all developers for a given app *must* use the `useQueryLoader` hook at all times. The intended usage is to only use it when you want the default behavior, according to how your app wants to define those defaults.

**Comment by FredyC** [comment]
Once again, how are you gonna solve when you need to do something else with data that requires other hooks? The default loading behavior is actually happening less often than one would think. Often we are using [react-content-loader](https://github.com/danilowoz/react-content-loader) to show svg specific to that page layout. When loading data to forms, we show the whole form with disabled fields until data are loaded.

Besides, loading spinners are less and less considered a good practice. User doesn't need to be distracted by some moving crap thing. In properly optimized apps when requests take a couple of ms, it's totally fine to just show nothing or some placeholder shape.

Doesn't matter. You can use what you like, but it's definitely impossible to expect everyone has similar needs.
